File: 백준/Gold/4803. 트리/트리.cc
#include <iostream>
#include <stack>
#include <algorithm>
#include <vector>
#include <queue>
using namespace std;

bool visited[501];
vector<int> v[501];



bool IsTree(int now, int level){
    visited[now] = 1;
    for(auto next : v[now]){
        if(next == level){
            continue;
        }
        if(visited[next]){
            
            return false;
        }
        if(IsTree(next, now) == false){
            //cout << now << ' ' << level << '\n';
            return false;
        }
    }
    return true;
}


int main(){
    cin.tie(0);
    ios_base::sync_with_stdio(false);

    int index = 1;
    while(1){
        int u, ver;
        cin >> u >> ver;
        if(u == 0 && ver == 0){
            break;
        }
        for(int i = 1; i <= u; i++){
            v[i].clear();
            visited[i] = 0;
        }

        for(int i = 0; i < ver; i++){
            int x1, x2; 
            cin >> x1 >> x2;
            v[x1].push_back(x2);
            v[x2].push_back(x1);
        }

        

        int cnt = 0;
        for(int i = 1; i <= u; i++){
            if(!visited[i]){
                if(IsTree(i, 0)){
                    cnt++;
                }
            }
        }

        if(cnt == 1){
            cout << "Case " << index << ": There is one tree." << '\n';
        }
        else if(cnt == 0){
            cout << "Case " << index << ": No trees." << '\n';
        }
        else{
            cout << "Case " << index << ": A forest of " << cnt << " trees." << '\n';
        }
        index++;
    }
    
    return 0;
}

File: 백준/Gold/4803. 트리/트리.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs")
  .readFileSync(filePath)
  .toString()
  .trim()
  .split("\n");

let index = 0;
let testCase = 1

function cycle(now, prev, visited, graph){
    visited[now] = 1;
    for(let y of graph[now]){
        if(!visited[y]){
            if(cycle(y, now, visited, graph)){
                return true;
            }
        }
        else if(y !== prev){
            return true;
        }
    }
    return false;
}

function solution(x, y){
    let visited = Array.from({length : x}, () => 0);
    let cnt = 0;
    let arr = Array.from({length : y});
    const graph = new Array(x + 1);
    for (let i = 1; i <= x; i += 1) graph[i] = [];

    for(let i = 0; i < y; i++){
        arr[i] = input[index++].split(' ').map(Number);
    }

    //console.log(arr);


    for(let i = 0; i < y; i++){
        const [x, y] = arr[i];
        //console.log(x, y);
        graph[x].push(y);
        graph[y].push(x);
    }

    for(let i = 1; i <= x; i++){
        if(!visited[i]){
            if(!cycle(i, 0, visited, graph)){
                cnt++;
            }
        }
    }
    
    if(cnt > 1){
        console.log(`Case ${testCase}: A forest of ${cnt} trees.`);
    }
    else if(cnt === 1){
        console.log(`Case ${testCase}: There is one tree.`);
    } else {
        console.log(`Case ${testCase}: No trees.`);
    }
    testCase += 1;
    
}

for(let i = 0; i < input.length; i++){
    let [n, m] = input[index++].split(' ').map(Number);
    if(n == 0){
        break;
    }
    solution(n, m);

}